File: langs/java/src/test/java/dev/dotlyte/MaskingTest.java
package dev.dotlyte;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for v2 masking / redaction.
 */
class MaskingTest {

    @Test
    void autoDetectSensitiveKeys() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("db_password", "secret123");
        data.put("api_key", "key-abc");
        data.put("host", "localhost");

        Set<String> sensitive = Masking.buildSensitiveSet(data, Collections.emptySet());
        assertTrue(sensitive.contains("db_password"));
        assertTrue(sensitive.contains("api_key"));
        assertFalse(sensitive.contains("host"));
    }

    @Test
    void redactMap() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("host", "localhost");
        data.put("password", "secret123");

        Set<String> sensitive = new HashSet<>(Arrays.asList("password"));
        Map<String, Object> redacted = Masking.redactMap(data, sensitive);

        assertEquals("localhost", redacted.get("host"));
        assertEquals(Masking.REDACTED, redacted.get("password"));
    }

    @Test
    void redactNestedMap() {
        Map<String, Object> inner = new LinkedHashMap<>();
        inner.put("host", "localhost");
        inner.put("password", "secret");

        Map<String, Object> data = new LinkedHashMap<>();
        data.put("db", inner);

        Set<String> sensitive = new HashSet<>(Arrays.asList("db.password"));
        Map<String, Object> redacted = Masking.redactMap(data, sensitive);

        @SuppressWarnings("unchecked")
        Map<String, Object> redactedDb = (Map<String, Object>) redacted.get("db");
        assertEquals("localhost", redactedDb.get("host"));
        assertEquals(Masking.REDACTED, redactedDb.get("password"));
    }

    @Test
    void formatRedacted() {
        assertEquals("se*********", Masking.formatRedacted("secret12345"));
        assertEquals("***", Masking.formatRedacted("abc"));
    }

    @Test
    void schemaKeysIncluded() {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("custom_field", "value");

        Set<String> schemaKeys = new HashSet<>(Arrays.asList("custom_field"));
        Set<String> sensitive = Masking.buildSensitiveSet(data, schemaKeys);
        assertTrue(sensitive.contains("custom_field"));
    }
}

File: langs/php/tests/MaskingTest.php
<?php

declare(strict_types=1);

namespace Dotlyte\Tests;

use Dotlyte\Masking;
use Dotlyte\SchemaRule;
use Dotlyte\Validator;
use PHPUnit\Framework\TestCase;

final class MaskingTest extends TestCase
{
    public function testAutoDetectSensitive(): void
    {
        $data = [
            'db_password' => 's3cret',
            'api_key' => 'abc123',
            'host' => 'localhost',
            'auth_token' => 'tok',
        ];
        $keys = Masking::buildSensitiveSet($data);
        $this->assertContains('db_password', $keys);
        $this->assertContains('api_key', $keys);
        $this->assertContains('auth_token', $keys);
        $this->assertNotContains('host', $keys);
    }

    public function testSchemaKeys(): void
    {
        $data = ['host' => 'localhost', 'secret' => 'abc'];
        $schema = [
            'secret' => new SchemaRule(type: 'string', sensitive: true),
        ];
        $schemaKeys = Validator::sensitiveKeys($schema);
        $keys = Masking::buildSensitiveSet($data, $schemaKeys);
        $this->assertContains('secret', $keys);
    }

    public function testRedact(): void
    {
        $data = ['host' => 'localhost', 'password' => 's3cret'];
        $result = Masking::redact($data, ['password']);
        $this->assertSame('localhost', $result['host']);
        $this->assertSame(Masking::REDACTED, $result['password']);
    }

    public function testNestedRedact(): void
    {
        $data = [
            'database' => ['host' => 'db', 'password' => 'secret'],
        ];
        $result = Masking::redact($data, ['database.password']);
        $this->assertSame('db', $result['database']['host']);
        $this->assertSame(Masking::REDACTED, $result['database']['password']);
    }

    public function testFormatRedacted(): void
    {
        $this->assertSame('ab***', Masking::formatRedacted('abcde'));
        $this->assertSame('**', Masking::formatRedacted('ab'));
        $this->assertSame(Masking::REDACTED, Masking::formatRedacted(null));
    }

    public function testRedactPreservesStructure(): void
    {
        $data = [
            'port' => 8080,
            'debug' => true,
            'api_key' => 'test-key-123',
        ];
        $result = Masking::redact($data, ['api_key']);
        $this->assertSame(8080, $result['port']);
        $this->assertTrue($result['debug']);
        $this->assertSame(Masking::REDACTED, $result['api_key']);
    }
}
